### DISCUSSION TITLE
Welcome to magritte Discussions!

### DISCUSSION THREAD

**Discussion by seandenigris** [magritte-metamodel/magritte#183] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by eMaringolo** [comment]
Thank you for setting this up @seandenigris!

Presentation? I've been using Magritte in production for quite a while now, and in 2006 at a company I worked in we forked Magritte (1.0?) to make our own version of it, adapted to work for the generation of UIs in Dolphin (using MVP), plus a few extensions we made, like creating an REST API based on the descriptors. Even when all of the classes were renamed to use a different prefix, it still felt like Magritte.

Ideas? I'd like a two-phase collection of descriptions, in Magritte 1.0 the descriptions were at class side, now it's at instance side, I'd like a mix of both, so you can have a generic list on the class side that doesn't require instantiating a possible invalid object just to get the list of available descriptions.

I think that using Magritte value grows as you make your system larger and/or more manipulable by the end user (queries, reports, etc.).

I hope we can have more discussions through this channel!

**Reply by seandenigris** [reply]
Yes, @eMaringolo - I was excited about this new GH feature! We have the Magritte ML, but the traffic is low and IMHO there is value in keeping discussion as close to the code as possible.

> adapted to work for the generation of UIs in Dolphin (using MVP)

I made the Morphic stuff functional (but still ugly). I always wanted to improve it but was waiting for Pharo to stabilize. I think it would probably be time to revisit once Spec2 is released. Did you use the visitor pattern for the UI stuff you did? I extended what was already there for Morphic (which didn't use the pattern), but I was always curious whether it should and did a promising experiment in this direction.

> forked Magritte (1.0?) to make our own version of it

I know that several people did there own work on top of Magritte, most notably QCMagritte. I'm very interested to see if things can be backported here so we can all benefit.

> I'd like a two-phase collection of descriptions...

That's an interesting idea. It's hard to have this be generally applicable because, now that descriptions are on the instance side, they can depend on the instance, but have blocks tied to whatever "self" was when they were created. I've thought about passing the instance as an optional last argument to all blocks held by descriptions. Maybe a code critic warning about "self in blocks" too. I guess in some cases, the per-instance flexibility is not actually used. One could envision descriptions created on the class-side and then fleshed out on the instance side, but I wonder if the complexity would be worth it...

> I think that using Magritte value grows as you make your system larger and/or more manipulable by the end user (queries, reports, etc.).

I agree. Although I feel I've gotten a lot of leverage just by getting Morphic forms for free. Even lazy init from descriptions is one less piece of boilerplate to write, multiplied by every field of every class I create.
```smalltalk
MyClass>>#name
	^ self maLazyInstVarUsing: self descriptionName
```

I need to do a documentation pass...

**Comment by sebastianconcept** [comment]
Hi @seandenigris great idea to start with this.

I never had Magritte in production so far but I remember using Magritte when setting up a wiki based on it circa 2002. I can't remember its name and neither Google or ChatGTP do either, sorry.

My current curiosity in Magritte is related to using it for validation in [Ride](https://github.com/sebastianconcept/ride) so I've opened this [issue](https://github.com/sebastianconcept/ride/issues/22) to explore using it in a way that feels easy from its Presenter classes aiming to be at least as productive as [RoR](https://guides.rubyonrails.org/active_record_validations.html) in this point and based on [Mapless](https://github.com/sebastianconcept/Mapless) instead of something like `ActiveRecord`.
